File: LeetCode/530. Minimum Absolute Difference in BST.py
# https://leetcode.com/problems/minimum-absolute-difference-in-bst/description/


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def getMinimumDifference(self, root: TreeNode | None) -> float:
        min_diff = float("inf")

        prev: TreeNode | None = None

        def dfs(node: TreeNode | None) -> None:
            if node is None:
                return

            dfs(node.left)

            nonlocal prev
            if prev is not None:
                nonlocal min_diff
                min_diff = min(min_diff, abs(prev.val - node.val))

            prev = node

            dfs(node.right)

        dfs(root)

        return min_diff

File: LeetCode/530. Minimum Absolute Difference in BST.js
// https://leetcode.com/problems/minimum-absolute-difference-in-bst/description/
const { test } = require('../test');

const getMinimumDifference = (node) => {
  let min = Number.POSITIVE_INFINITY;

  let prev = null;

  const dfs = (node) => {
    if (!node) return;

    if (prev) {
      min = Math.min(min, Math.abs(prev.val - node.val));
    }

    prev = node;

    dfs(node.left);
    prev = node;
    dfs(node.right);
  };

  dfs(node);

  return min;
};
